libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -475,66 +598,128 @@ genFreshCredential tries0 tag old = go tries0
+        then genFreshKeyHash $ n - 1
+        else return kh
+
+-- | Use this function to get a new pool that should not be used in the future transactions
+genNewPool :: forall era. Reflect era => GenRS era (KeyHash 'StakePool (Crypto era), PoolParams (Crypto era), IndividualPoolStake (Crypto era))
+genNewPool = do
+  poolId <- genFreshKeyHash 100
+  poolparam <- genPoolParams poolId
+  percent <- lift $ choose (0, 1 :: Float)
+  let stake = IndividualPoolStake @(Crypto era) (toRational percent) (_poolVrf poolparam)
+  modify (\s -> s {gsAvoidKey = Set.insert (coerceKeyRole poolId) $ gsAvoidKey s})
+  pure (poolId, poolparam, stake)
+
+addPoolToInitialState ::
+  KeyHash 'StakePool (Crypto era) ->
+  PoolParams (Crypto era) ->
+  IndividualPoolStake (Crypto era) ->
+  GenRS era ()
+addPoolToInitialState poolId poolparam stake =
+  modify
+    ( \st ->
+        st
+          { gsInitialPoolParams = Map.insert poolId poolparam (gsInitialPoolParams st),
+            gsInitialPoolDistr = Map.insert poolId stake (gsInitialPoolDistr st)
+          }
+    )
+
+addPoolToModel :: KeyHash 'StakePool (Crypto era) -> PoolParams (Crypto era) -> IndividualPoolStake (Crypto era) -> GenRS era ()
+addPoolToModel poolId poolparam stake =
+  modifyModel
+    ( \m ->
+        m
+          { mPoolParams = Map.insert poolId poolparam (mPoolParams m),
+            mPoolDistr = Map.insert poolId stake (mPoolDistr m)
+          }
+    )
+
+genPoolParams ::
+  Reflect era =>
+  KeyHash 'StakePool (Crypto era) ->
+  GenRS era (PoolParams (Crypto era))
+genPoolParams _poolId = do
+  _poolVrf <- lift arbitrary
+  _poolPledge <- lift genPositiveVal
+  _poolCost <- lift genPositiveVal
+  _poolMargin <- lift arbitrary
+  _poolRAcnt <- RewardAcnt Testnet <$> genCredential Rewrd
+  let _poolOwners = mempty
+  let _poolRelays = mempty
+  let _poolMD = SNothing
+  pure PoolParams {..}
 
 -- Adds to both gsKeys and gsScripts and gsPlutusScript
 -- via genKeyHash and genScript
 
 -- | Generate a credential that can be used for supplied purpose (in case of
 -- plutus scripts), while occasionally picking out randomly from previously
 -- generated set.
+-- Returns the credential and True iff the credential is freshly generated
 genCredential :: forall era kr. Reflect era => Tag -> GenRS era (Credential kr (Crypto era))
 genCredential tag =
   frequencyT
-    [ (35, KeyHashObj <$> genKeyHash),
-      (35, ScriptHashObj <$> genScript reify tag),
+    [ (35, KeyHashObj <$> genKeyHash'),
+      (35, ScriptHashObj <$> genScript' (100 :: Int)),
       (10, pickExistingKeyHash),
       (20, pickExistingScript)
     ]
   where
+    genKeyHash' = do
+      kh <- genKeyHash
+      case tag of
+        Rewrd -> modify $ \st ->
+          st {gsInitialRewards = Map.insert (KeyHashObj kh) (Coin 0) $ gsInitialRewards st}
+        _ -> pure ()
+      return $ coerceKeyRole kh
+    genScript' n
+      | n <= 0 = error "Failed to generate a fresh script hash"
+      | otherwise = do
```

Comment:
Same thing here, we don't need the `n` argument. I understand that this is already a local binding, but accepting an argument makes the reader question what does this argument mean and why an arbitrary number of 100 was used.